Aula de vim (#16)
guia-de-sobrevivencia-do-programador/Notas-de-Aula: 3 - Editores (VIM, EMACS, VSCode)/vim.md

# Vi , Vim e Neovim

Vim é um editor antigo, de código fonte aberto, muito popular entre programadores por ser altamente extensível, e infame por sua curva de aprendizado não amigável a iniciantes. Entretanto, ele é uma versão melhorada (*v* *im*proved) do editor `vi`, que não continha _syntax highlighting_, modo visual, scripting e outras gostosuras.

O vi, predecessor do vim, foi concebido nos anos 70 por Billy Joy com base no editor `ex`, e tem seu nome por ser o modo `*vi*sual` deste. O vi surgiu em uma época onde se acessava computadores através de terminais remotos cujo layout dos teclados era levemente diferente: As setas direcionais eram acessadas pelas teclas `hjkl` e a tecla Escape econtrava-se onde hoje encontramos a tecla Caps Lock.

<img title="teclado do terminal ADM3A" alt="teclado antigo kk" src="./Resources/ADM3A.png">

> O vim por sua vez foi criado e mantido por Bram Moolenaar, porém por ser ainda o único *maintainer* do repositório (não permitindo contribuições no código por ninguém além dele mesmo), as atualizações do editor não conseguem acompanhar as vontades da comunidade e, devido a isso, foi criado o editor Neovim, que é um *fork* do editor vim, visando refatorar o codigo do editor e adicionar novas funcionalidades de maneira mais rápida e centradas na comunidade, mas ainda mantendo compatibilidade com o seu predecessor. Das várias contribuições feitas pela comunidade voltadas para o usuário final, duas que se destacam são o suporte nativo para Language Server Protocols e a adoção de Lua como sua linguagem principal para configuracao e scripts.

Reviewer: ```suggestion
O vim por sua vez foi criado e mantido por Bram Moolenaar, porém por ser ainda o único *maintainer* do repositório (não permitindo contribuições no código por ninguém além dele mesmo), as atualizações do editor não conseguem acompanhar as vontades da comunidade e, devido a isso, foi criado o editor Neovim, que é um *fork* do editor vim, visando refatorar o código do editor e adicionar novas funcionalidades de maneira mais rápida e centradas na comunidade, mas ainda mantendo compatibilidade com o seu predecessor. Das várias contribuições feitas pela comunidade voltadas para o usuário final, duas que se destacam são o suporte nativo para Language Server Protocols e a adoção de Lua como sua linguagem principal para configuração e scripts.
```

# Por quê VIM?

Vim preza por uma edicao fluida de codigo. O modo de funcionamento padrao do vim e o `normal mode`. Nesse modo, cada tecla tem um significado e o que voce faz e conversar com o editor atraves de frases.

> Essas frases tem uma acao (com um numero de repeticoes como argumento opcional) e, dependendo da acao, um objeto ou movimento sob o qual essa acao sera executada. Por um lado, uma acao pode ser `delete`, `replace`, `change`, `insert`, `paste` e muitas outras. Por outro lado, um objeto ou movimento pode ser uma ou mais (através de argumentos opcionais) palavras, parágrafos, linhas, tudo dentro de um parentêses e várias outras coisas também.

Reviewer: ```suggestion
Essas frases tem uma ação (com um numero de repetições como argumento opcional) e, dependendo da ação, um objeto ou movimento sob o qual essa ação será executada. Por um lado, uma ação pode ser `delete`, `replace`, `change`, `insert`, `paste` e muitas outras. Por outro lado, um objeto ou movimento pode ser uma ou mais (através de argumentos opcionais) palavras, parágrafos, linhas, tudo dentro de um parênteses e várias outras coisas também.
```

## Como abrir um arquivo:
Para abrir um arquivo com vim, basta digitarmos no terminal: `vim [nome-do-arquivo]`. Caso o arquivo não exista, o editor tratará de criar esse arquivo para você.

## Ué, porque o que eu escrevo não aparece na tela?

Se você acabou de abrir o editor e já tentou escrever alguma coisa, vai perceber que o resultado não foi nada esperado. 
Isso se dá porque vim funciona através de modos: Quando abrimos o editor, ele está no modo normal, que permite a navegação pelo texto, mas não a edição.
Para editarmos, basta digitar `i` dentro do modo normal, o que nos levará ao modo de inserção, nele podemos digitar sem obstáculos. 
Para voltar ao modo normal, basta apertar a tecla Esc. Essa lógica vale para (quase) todos os modos que você se encontrar.

## Ok, mas como que eu salvo um arquivo?
Você não sai. MUAHAHAHAHAHAH. Brincadeira. Para salvarmos um arquivo, precisamos dizer ao editor que vamos executar um comando.
Para isso, precisamos estar no modo normal e então digitar `:w` e apertar a tecla Enter. Os dois pontos ativam o modo de linha de comando, 
já o caractere `w` (de _write_) indica o comando de salvar o conteúdo escrito na tela para o arquivo. 

> A partir de agora é só colinha para refrescar os comandos (ATENÇãO! `^` antes de um caractere significa que a tecla CTRL deve ser pressionada junto ao mesmo):

Reviewer: ```suggestion
A partir de agora é só colinha para refrescar os comandos (ATENÇÃO! `^` antes de um caractere significa que a tecla CTRL deve ser pressionada junto ao mesmo):
```

> # vi , vim e Neovim

Reviewer 2: ```suggestion
# Vi , Vim e Neovim
```

Vim é um editor de código fonte aberto antigo, muito popular entre programadores por ser altamente extensível e infame por sua dificuldade inicial de uso.

> Entretanto, vim é uma versão melhorada (*vi* i*m*proved) do editor `vi`, que não continha _syntax highlighting_, modo visual, scripting e outras gostosuras.

Reviewer 2: N esquece de coloca acentos (eu n tenho no meu teclado e to com preguica de colocar)...

```suggestion
Vim é um editor antigo, de código fonte aberto, muito popular entre programadores por ser altamente extensível, e infame por sua curva de aprendizado nao amigavel a iniciantes. Entretanto, ele é uma versão melhorada (*v* *im*proved) do editor `vi`, que não continha _syntax highlighting_, modo visual, scripting e outras gostosuras.

```

Billy Joy programou o vi nos anos 70,  com base em no editor `ex` e tem seu nome por ser o modo `*vi*sual` deste. O vi surgiu em uma época onde se acessava computadores 
através de terminais remotos, cujo layout dos teclados era levemente diferente: As setas direcionais eram acessadas pelas teclas `hjkl` e a tecla Escape 

> econtrava-se onde encontramos a Caps Lock.

Reviewer 2: ```suggestion
O vi, predecessor do vim, foi concebido nos anos 70 por Billy Joy com base no editor `ex`, e tem seu nome por ser o modo `*vi*sual` deste. O vi surgiu em uma época onde se acessava computadores através de terminais remotos cujo layout dos teclados era levemente diferente: As setas direcionais eram acessadas pelas teclas `hjkl` e a tecla Escape econtrava-se onde hoje encontramos a tecla Caps Lock.
```

<img title="teclado do terminal ADM3A" alt="teclado antigo kk" src="./Resources/ADM3A.png">

Vim foi criado e mantido por Bram Moolenaar, porém o fato de ser ainda o único *maintainer* do repositório, as atualizações do editor não conseguem acompanhar 
as vontades da comunidade e devido a isso, criou-se o editor Neovim, que é um *fork* do editor vim, visando refatorar o editor e adicionar novas funcionalidades 
de maneira mais rápida, mas ainda mantendo compatibilidade com o seu predecessor. Neovim tem suporte nativo a Language Server Protocols, adota Lua como sua linguagem principal para

> scripts [ETC ETC PREENCHER]

Reviewer 2: Acentuacao...

```suggestion
O vim por sua vez foi criado e mantido por Bram Moolenaar, porém por ser ainda o único *maintainer* do repositório (nao permitindo contribuicoes no codigo de ninguem alem dele mesmo), as atualizações do editor não conseguem acompanhar as vontades da comunidade e, devido a isso, foi criado o editor Neovim, que é um *fork* do editor vim, visando refatorar o codigo do editor e adicionar novas funcionalidades de maneira mais rápida e centradas na comunidade, mas ainda mantendo compatibilidade com o seu predecessor. Das varias contribuicoes feitas pela comunidade voltadas para o usuario final, duas que se destacam sao o suporte nativo para Language Server Protocols e a adocao de Lua como sua linguagem principal para configuracao e scripts.
```

> # Por quê VIM?

Reviewer 2: O que ta entre parenteses eu nao sei se precisa

```suggestion
# Por quê VIM?

Vim preza por uma edicao fluida de codigo. O modo de funcionamento padrao do vim e o `normal mode`. Nesse modo, cada tecla tem um significado e o que voce faz e conversar com o editor atraves de frases.

Essas frases tem uma acao (com um numero de repeticoes como argumento opcional) e, dependendo da acao, um objeto ou movimento sob o qual essa acao sera executada. Por um lado, uma acao pode ser `delete`, `replace`, `change`, `insert`, `paste` e muitas outras. Por outro lado, um objeto ou movimento pode ser uma ou mais (atraves de argumentos opcionais) palavras, paragrafos, linhas, tudo dentro de um parenteses e varias outras coisas tambem.

Essa gramatica, quando dominada, permite uma edicao muito mais rapida e fluida do codigo sem o sentimento de que voce esta pensando para programar. Voce pode somente focar no codigo em si e nao no ato de programar.

```

> Para os acostumados com a sintaxe e o jeito de funcionar do vim, editar um arquivo se torna uma torna muito menos monótona e muito mais dinâmica: 

Reviewer 2: ```suggestion
Para os acostumados com a sintaxe e o jeito de funcionar do vim, editar um arquivo se torna uma tarefa muito menos monótona e muito mais dinâmica: 
```

> Ações complexas requerem poucas teclas digitadas para serem computadas e tarefas repetitivas podem ser facilmente repetidas com uma ou duas teclas (`.` ou `@`).

Reviewer 2: Deixa para falar de teclas quando comecar o "tutorial", n taca coisas que vc vai ter que explicar mais do que ja foi explicado so para ter que explicar de novo depois.

```suggestion
Ações complexas requerem poucas teclas digitadas para serem executadas, e tarefas repetitivas podem ser facilmente reexecutadas em outro trecho de codigo ou ate em varios arquivos diferentes.
```

> Adicionalmente, devido a sua origem, o editor força o usuário a nunca tirar a mão de perto das letras, e isso permite que ações sejam encadeadas sem interrupções. 

Reviewer 2: ```suggestion
Adicionalmente, devido a sua origem, o editor força o usuário a nunca tirar a mão de perto das letras (as chamadas _home row keys_), e isso permite que ações sejam encadeadas sem interrupções. 
```

> Além disso, vim permite que funções novas de edição sejam criadas, dando liberdade ao usuário de customizar, através do arquivo .vimrc.

Reviewer 2: ```suggestion
Além disso, vim permite que funções novas de edição sejam criadas, dando liberdade ao usuário de customizar qualquer aspecto do editor através de scripts feitos por voce e plugins feitos pela comunidade.
```

# Os básicos.

## Como abrir um arquivo:
Para abrir um arquivo com vim, basta digitarmos no terminal: `vim [nome-do-arquivo]`. Caso o arquivo não exista, o editor tratará de criar esse arquivo para você.

## Ué, porque o que eu escrevo não aparece na tela?

Se você acabou de abrir o editor e já tentou escrever alguma coisa, vai perceber que o resultado não foi nada esperado. 
Isso se dá porque vim funciona através de modos: Quando abrimos o editor, ele está no modo normal, que permite a navegação pelo texto, mas não a edição.
Para editarmos, basta digitar `i` dentro do modo normal, o que nos levará ao modo de inserção, nele podemos digitar sem obstáculos. 

> Para sairmos do modo de inserção, basta apertar a tecla Esc que voltaremos ao modo normal. Essa lógica vale para todos os modos que você se encontrar.

Reviewer 2: ```suggestion
Para voltar ao modo normal, basta apertar a tecla Esc. Essa lógica vale para (quase) todos os modos que você se encontrar.
```

## Ok, mas como que eu salvo um arquivo?
Você não sai. MUAHAHAHAHAHAH. Brincadeira. Para salvarmos um arquivo, precisamos dizer ao editor que vamos executar um comando.

> Para isso, precisamos estar no modo normal e então digitar `:w` e apertar a tecl Enter. Os dois pontos ativam o modo de linha de comando, 

Reviewer 2: ```suggestion
Para isso, precisamos estar no modo normal e então digitar `:w` e apertar a tecla Enter. Os dois pontos ativam o modo de linha de comando, 
```

> já o caractere `w` indica o comando de salvar o conteúdo escrito na tela para o arquivo. 

Reviewer 2: ```suggestion
já o caractere `w` (de _write_) indica o comando de salvar o conteúdo escrito na tela para o arquivo. 
```

A partir de agora é só colinha para refrescar os comandos:

# Modo Normal
## Movimentação

-  Toda movimentação pode receber uma quantidade antes que indica quantas vezes será repetida. Ex: `4j` ou `50l`
-  Por caracteres: `h`(esquerda), `j`(baixo), `k`(cima), `l`(direita)

> -  Por "palavra": `w`(começo da próxima palavra, `e`(final da próxima palavra), `b` (começo da palavra anterior)

Reviewer 2: ```suggestion
-  Por "palavra": `w` (`word`: começo da próxima palavra), `e`(`end of word`: final da próxima palavra), `b` (`back`: começo da palavra anterior)
```

> -  Por linha (não aceita quantidade): `0`(começo da linha), `$`(final da linha)

Reviewer 2: ```suggestion
-  Por linha (não aceita quantidade): `0`(começo da linha), `^`(comeco da linha ignorando caracteres de espaco), `$`(final da linha)
```

Author: Boa, eu uso toda hora, mas fiz remap então acabo esquecendo

> -  Por arquivo: `gg`(Começo do arquivo), `G`(Final do arquivo) 

Reviewer 2: ```suggestion
-  Por arquivo: `gg`(Começo do arquivo), `G`(Final do arquivo)
-  Por arquivo (tambem): `#G`(substitua `#` por um numero: va para a linha `#`), `#%`(substitua `#` por um numero: va para o `#` por cento do arquivo. Ex.: `50%`: va para o meio do arquivo)
```

Reviewer 2: Sao divertidos mas n sei se precisa colocar, vc decide se tem tempo ou n

Author: Vou colocar: estar nas notas de aula não mata. Mas realmente talvez não caiba na aula.

## Desfazer ações
-  Desfazer: `u`

> -  Desfazer o desfazer: `R`

Reviewer 2: ```suggestion
-  Refazer: `R`
```

Reviewer 2: E nao e `^r` em vez de `R`?

Author: É sim, eu deixei passar batido

## Deletar
-  Aceita quantidades antes da ação
-  Deletar por movimento: `d` + movimento
-  Deletar linha inteira: `dd`
-  Deletar até o final da linha: `D`

## Copiar e Colar
-  Copiar por movimento: `y` + movimento

> -  Copiar linha inteira: `Y` ou `yy`

Reviewer 2: `Y` copia a partir do seu cursor. Se ele nao estiver no comeco da linha vc n copia a linha inteira (diferente do `yy`).
Mesma coisa que `dd` e `D`.

Author: Interessante que no meu vim isso não acontece e não me lembro de ter feito remap disso.
Estranho

Author: Na docs do vim fala que o Y como `y$`  não é o comportamento padrão

-  Colar antes do cursor: `P`
-  Colar depois do cursor: `p`

> # Modo de Inserção (entra-se a partir do modo normal)

Reviewer 2: Sera que seria legal colocar tambem `A` e `I`?

Author: Posso colocar como referência mas talvez estoure o tempo na aula

-  Inserir antes do cursor: `i`
-  Adicionar ao final do cursor: `a` 
-  Sair do modo normal : Tecla Esc

# Modo visual (entra-se a partir do modo normal)
-  Funcionalidade igual ao modo normal, porém selecionando texto.
-  Ativar modo visual por caractere: `v`
-  Ativar modo visual por linha: `V`
-  Ativar modo visual por coluna: `CTRL+V`

> # Modo de comandos (entra-se a partir do modo normal ou visual)

Reviewer 2: Coloca a opcao `:%s/adsf/asdf/` tambem. O `%` indica "no arquivo inteiro" em vez de "nessa linha"

Author: Então, isso eu quis omitir por causa do jeito que eu ia ensinar, que é fazendo a seleção pelo modo visual e então fazendo o search and replace. Mas acho melhor colocar também

-  Ativar modo:  `:`
-  Procurar e substituir uma vez na linha: `s/PADRAO-DE-PROCURA/TEXTO-DE-SUBSTITUICAO/` + tecla ENTER
-  Procurar e substituir todas as ocorrências na linha: `s/PADRAO-DE-PROCURA/TEXTO-DE-SUBSTITUICAO/g` + tecla ENTER
-  Procurar e substituir confirmando a substituição: `s/PADRAO-DE-PROCURA/TEXTO-DE-SUBSTITUICAO/c` + tecla ENTER

> # BIBLIOGRAFIA e extras

Reviewer 2: Ve a bibliografia da aula 2 e tenta seguir o padrao (ou faz um PR para mudar a bibliografia da aula 2 para ficar igual)

Author: Vou alterar esse daqui mesmo

# Vi , Vim e Neovim

> Vim é um editor antigo, de código fonte aberto, muito popular entre programadores por ser altamente extensível, e infame por sua curva de aprendizado nao amigavel a iniciantes. Entretanto, ele é uma versão melhorada (*v* *im*proved) do editor `vi`, que não continha _syntax highlighting_, modo visual, scripting e outras gostosuras.

Reviewer: ```suggestion
Vim é um editor antigo, de código fonte aberto, muito popular entre programadores por ser altamente extensível, e infame por sua curva de aprendizado não amigável para iniciantes. Entretanto, ele é uma versão melhorada (*v* *im*proved) do editor `vi`, que não continha _syntax highlighting_, modo visual, scripting e outras gostosuras.
```

Gostei do "gostosuras". :-D

Author: Feito em 9c10d44de7ead3a26910edaed66609cbc7033535

O vi, predecessor do vim, foi concebido nos anos 70 por Billy Joy com base no editor `ex`, e tem seu nome por ser o modo `*vi*sual` deste. O vi surgiu em uma época onde se acessava computadores através de terminais remotos cujo layout dos teclados era levemente diferente: As setas direcionais eram acessadas pelas teclas `hjkl` e a tecla Escape econtrava-se onde hoje encontramos a tecla Caps Lock.

<img title="teclado do terminal ADM3A" alt="teclado antigo kk" src="./Resources/ADM3A.png">

> O vim por sua vez foi criado e mantido por Bram Moolenaar, porém por ser ainda o único *maintainer* do repositório (nao permitindo contribuicoes no codigo de ninguem alem dele mesmo), as atualizações do editor não conseguem acompanhar as vontades da comunidade e, devido a isso, foi criado o editor Neovim, que é um *fork* do editor vim, visando refatorar o codigo do editor e adicionar novas funcionalidades de maneira mais rápida e centradas na comunidade, mas ainda mantendo compatibilidade com o seu predecessor. Das varias contribuicoes feitas pela comunidade voltadas para o usuario final, duas que se destacam sao o suporte nativo para Language Server Protocols e a adocao de Lua como sua linguagem principal para configuracao e scripts.

Reviewer: ```suggestion
O vim por sua vez foi criado e mantido por Bram Moolenaar, porém por ser ainda o único *maintainer* do repositório (não permitindo contribuições no código de ninguém além dele mesmo), as atualizações do editor não conseguem acompanhar as vontades da comunidade e, devido a isso, foi criado o editor Neovim, que é um *fork* do editor vim, visando refatorar o código do editor e adicionar novas funcionalidades de maneira mais rápida e centradas na comunidade, mas ainda mantendo compatibilidade com o seu predecessor. Das varias contribuições feitas pela comunidade voltadas para o usuário final, duas que se destacam são o suporte nativo para Language Server Protocols e a adoção de Lua como sua linguagem principal para configuração e scripts.
```

# Por quê VIM?

> Vim preza por uma edicao fluida de codigo. O modo de funcionamento padrao do vim e o `normal mode`. Nesse modo, cada tecla tem um significado e o que voce faz e conversar com o editor atraves de frases.

Reviewer: ```suggestion
Vim preza por uma edição fluida de código. O modo de funcionamento padrão do vim e o `normal mode`. Nesse modo, cada tecla tem um significado e o que você faz e conversar com o editor através de frases.
```
